src/mscorlib/corefx/System/Runtime/InteropServices/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Allows limited thread safe reuse of heap buffers to limit memory pressure.
+    /// 
+    /// This cache does not ensure that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
```

Comment:
> What isn't available in CoreRT?

SafeBuffer is not available in CoreRT CoreLib because of it is high level interop type. High level interop is in System.Private.Interop in CoreRT, and I do not see a good reason to change it.

> The ability to resize in place and retain content.

This is opportunistic optimization. The equivalent opportunistic optimization with ArrayPool is that it will give you bigger buffer than what you have asked for.

> Lock free caching.

Array pool has that already.


The problem with these local adhoc pool is that they look great in microbenchmarks, but they hurt in real workloads: wasting memory because of you keep the pooled block forever even if it is needed just once in a blue moon, trashing cache (e.g. imagine what happens when multiple cores are operating on paths - you will be silently wasting a lot of cycles on memory bus contention as the different cores will be fighting over the one static cache). Now, the ArrayPool has some of these problems as well (e.g. the retention policy for large buffers), but it is something we can reason about and fix once it is used in more places and we can see the effects of the tuning on real workloads.